File: sdk/typescript/src/exceptions.ts
/**
 * Exception classes for Loopai SDK
 */

/**
 * Base exception for all Loopai SDK errors
 */
export class LoopaiException extends Error {
  statusCode?: number;
  response?: unknown;

  constructor(message: string, statusCode?: number, response?: unknown) {
    super(message);
    this.name = 'LoopaiException';
    this.statusCode = statusCode;
    this.response = response;
    Object.setPrototypeOf(this, LoopaiException.prototype);
  }
}

/**
 * Exception raised when input or output validation fails
 */
export class ValidationException extends LoopaiException {
  errors: Record<string, string[]>;

  constructor(
    message: string,
    errors: Record<string, string[]> = {},
    statusCode: number = 400
  ) {
    super(message, statusCode);
    this.name = 'ValidationException';
    this.errors = errors;
    Object.setPrototypeOf(this, ValidationException.prototype);
  }
}

/**
 * Exception raised when program execution fails
 */
export class ExecutionException extends LoopaiException {
  executionId?: string;

  constructor(message: string, executionId?: string, statusCode: number = 500) {
    super(message, statusCode);
    this.name = 'ExecutionException';
    this.executionId = executionId;
    Object.setPrototypeOf(this, ExecutionException.prototype);
  }
}

/**
 * Exception raised when connection to Loopai API fails
 */
export class ConnectionException extends LoopaiException {
  originalError?: Error;

  constructor(message: string, originalError?: Error) {
    super(message);
    this.name = 'ConnectionException';
    this.originalError = originalError;
    Object.setPrototypeOf(this, ConnectionException.prototype);
  }
}

File: sdk/python/loopai/exceptions.py
"""Exception classes for Loopai SDK."""

from typing import Any, Dict, Optional


class LoopaiException(Exception):
    """Base exception for all Loopai SDK errors."""

    def __init__(
        self,
        message: str,
        status_code: Optional[int] = None,
        response: Optional[Dict[str, Any]] = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.response = response

    def __str__(self) -> str:
        if self.status_code:
            return f"HTTP {self.status_code}: {self.message}"
        return self.message


class ValidationException(LoopaiException):
    """Exception raised when input or output validation fails."""

    def __init__(
        self,
        message: str,
        status_code: int = 400,
        errors: Optional[Dict[str, list[str]]] = None,
    ) -> None:
        super().__init__(message, status_code)
        self.errors = errors or {}

    def __str__(self) -> str:
        if self.errors:
            error_details = ", ".join(
                f"{field}: {', '.join(msgs)}" for field, msgs in self.errors.items()
            )
            return f"{self.message} - {error_details}"
        return self.message


class ExecutionException(LoopaiException):
    """Exception raised when program execution fails."""

    def __init__(
        self,
        message: str,
        status_code: int = 500,
        execution_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, status_code)
        self.execution_id = execution_id

    def __str__(self) -> str:
        if self.execution_id:
            return f"Execution {self.execution_id} failed: {self.message}"
        return f"Execution failed: {self.message}"


class ConnectionException(LoopaiException):
    """Exception raised when connection to Loopai API fails."""

    def __init__(self, message: str, original_exception: Optional[Exception] = None) -> None:
        super().__init__(message)
        self.original_exception = original_exception

    def __str__(self) -> str:
        if self.original_exception:
            return f"{self.message} ({type(self.original_exception).__name__}: {self.original_exception})"
        return self.message
